File: frontend/src/types/annotation.ts
export type AnnotationType = 'classification' | 'bbox' | 'polygon' | 'keypoint'

export interface Annotation {
  id: number
  task_image_id: number
  label_class_id: number | null
  annotation_type: AnnotationType
  data: Record<string, number | number[]>
  created_at: string
  updated_at: string
}

export interface AnnotationCreate {
  label_class_id?: number | null
  annotation_type: AnnotationType
  data: Record<string, number | number[]>
}

File: backend/app/schemas/annotation.py
from datetime import datetime

from pydantic import BaseModel, Field


class AnnotationCreate(BaseModel):
    label_class_id: int | None = None
    annotation_type: str = Field(min_length=1, max_length=20)
    data: dict = Field(default_factory=dict)


class AnnotationUpdate(BaseModel):
    label_class_id: int | None = None
    data: dict | None = None


class AnnotationResponse(BaseModel):
    id: int
    task_image_id: int
    label_class_id: int | None
    annotation_type: str
    data: dict
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BulkSaveRequest(BaseModel):
    annotations: list[AnnotationCreate]


class BulkSaveResponse(BaseModel):
    annotations: list[AnnotationResponse]
